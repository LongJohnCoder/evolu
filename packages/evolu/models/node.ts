import nanoid from 'nanoid';
import { IO } from 'fp-ts/lib/IO';
import { Newtype, iso } from 'newtype-ts';

/**
 * NodeID is string created with nanoid.
 */
export interface NodeID
  extends Newtype<{ readonly NodeID: unique symbol }, string> {}

const isoNodeID = iso<NodeID>();

export const mapNodeIDToString = (nodeID: NodeID): string =>
  isoNodeID.unwrap(nodeID);

export const id: IO<NodeID> = () => isoNodeID.wrap(nanoid());

export interface Node {
  readonly id: NodeID;
}
