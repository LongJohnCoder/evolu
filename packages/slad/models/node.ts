import { Brand } from 'utility-types';
import nanoid from 'nanoid';

/**
 * EditorNodeID is string created with nanoid.
 * Editor nodes need UUID for CRDT, React keys, and the other identity operations.
 */
export type EditorNodeID = Brand<string, 'EditorNodeID'>;

export interface EditorNode {
  readonly id: EditorNodeID;
}

export function isEditorNode(value: unknown): value is EditorNode {
  return value != null && typeof (value as EditorNode).id === 'string';
}

export function id(): EditorNodeID {
  return nanoid() as EditorNodeID;
}
