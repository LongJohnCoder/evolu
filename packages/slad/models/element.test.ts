import {
  normalizeEditorElement,
  EditorElement,
  isNormalizedEditorElement,
} from './element';
import { EditorNodeID } from './node';

let lastID = 0;

// Stable EditorNodeID factory for test snapshots.
function id(): EditorNodeID {
  return (lastID++).toString() as EditorNodeID;
}

test('normalizeEditorElement merges adjacent strings', () => {
  const element: EditorElement = {
    id: id(),
    children: [
      { id: id(), text: 'a' },
      { id: id(), text: 'b' },
      {
        id: id(),
        children: [
          { id: id(), text: 'a' },
          { id: id(), text: 'b' },
          { id: id(), text: 'c' },
          {
            id: id(),
            children: [
              { id: id(), text: 'a' },
              { id: id(), children: [] },
              { id: id(), text: 'b' },
              { id: id(), text: 'c' },
            ],
          },
        ],
      },
      { id: id(), text: 'a' },
      { id: id(), text: '' },
      { id: id(), text: 'b' },
    ],
  };
  expect(normalizeEditorElement(element)).toMatchSnapshot();
});

test('isNormalizedEditorElement', () => {
  expect(
    isNormalizedEditorElement({
      id: id(),
      children: [{ id: id(), text: 'a' }],
    }),
  ).toBe(true);
  expect(
    isNormalizedEditorElement({
      id: id(),
      children: [{ id: id(), text: '' }],
    }),
  ).toBe(true);
  expect(
    isNormalizedEditorElement({
      id: id(),
      children: [{ id: id(), text: 'a' }, { id: id(), text: 'b' }],
    }),
  ).toBe(false);
  expect(
    isNormalizedEditorElement({
      id: id(),
      children: [{ id: id(), children: [{ id: id(), text: '' }] }],
    }),
  ).toBe(true);
});

test('normalizeEditorElement do not add children', () => {
  expect(normalizeEditorElement({ id: id(), children: [] })).toMatchSnapshot();
});

test('normalizeEditorElement do not remove children', () => {
  expect(normalizeEditorElement({ id: id(), children: [] })).toMatchSnapshot();
  expect(
    normalizeEditorElement({ id: id(), children: [{ id: id(), text: '' }] }),
  ).toMatchSnapshot();
  expect(
    normalizeEditorElement({ id: id(), children: [{ id: id(), text: '.' }] }),
  ).toMatchSnapshot();
});
