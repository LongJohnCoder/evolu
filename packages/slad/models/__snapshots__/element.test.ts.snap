// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalizeEditorElement do not add children 1`] = `
Object {
  "children": Array [],
  "id": "29",
}
`;

exports[`normalizeEditorElement do not remove children 1`] = `
Object {
  "children": Array [],
  "id": "30",
}
`;

exports[`normalizeEditorElement do not remove children 2`] = `
Object {
  "children": Array [
    Object {
      "id": "32",
      "text": "",
    },
  ],
  "id": "31",
}
`;

exports[`normalizeEditorElement do not remove children 3`] = `
Object {
  "children": Array [
    Object {
      "id": "34",
      "text": ".",
    },
  ],
  "id": "33",
}
`;

exports[`normalizeEditorElement merges adjacent strings 1`] = `
Object {
  "children": Array [
    Object {
      "id": "1",
      "text": "ab",
    },
    Object {
      "children": Array [
        Object {
          "id": "4",
          "text": "abc",
        },
        Object {
          "children": Array [
            Object {
              "id": "8",
              "text": "a",
            },
            Object {
              "children": Array [],
              "id": "9",
            },
            Object {
              "id": "10",
              "text": "bc",
            },
          ],
          "id": "7",
        },
      ],
      "id": "3",
    },
    Object {
      "id": "12",
      "text": "a",
    },
    Object {
      "id": "13",
      "text": "",
    },
    Object {
      "id": "14",
      "text": "b",
    },
  ],
  "id": "0",
}
`;
